### ISSUE TITLE
BN - Batch Normalization

### ISSUE THREAD

**Issue by junxnone** [junxnone/tio#19]
BN - Batch Normalization - 批量标准化

- BN 的作用就是将输入值进行标准化，降低 scale 的差异至同一个范围内
  - 减少了前面一层参数的变化对后面一层输入值的影响，每一层独立训练
- 提高梯度的收敛程度
- 加快训练速度
- 可以使用较大的学习率
- 允许在深层网络中使用sigmoid这种易导致梯度消失的激活函数
- 具有轻微地正则化效果，以此可以降低dropout的使用

# Reference
- [ ] [深入理解Batch Normalization批标准化](https://www.cnblogs.com/guoyaohua/p/8724433.html)
- [ ] [深度学习基础系列（七）| Batch Normalization](http://www.cnblogs.com/hutao722/p/9842199.html)
- [ ] [深度学习中 Batch Normalization为什么效果好？](https://www.zhihu.com/question/38102762)
- [ ] [paper - Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift](https://arxiv.org/pdf/1502.03167.pdf)
- [ ] [Batch Normalization论文解读与Inception V2代码简析](https://blog.csdn.net/stesha_chen/article/details/81662405)
- [ ] [keras - BatchNormalization layer](https://keras.io/layers/normalization/)

# Brief
**Internal Covariate Shift**
-  训练深度神经网络的过程中，前一层参数改变了，后一层输入的分布也改变了
-  这需要比较小的学习率和特殊的参数初始化，进而导致训练速度变慢，有可能不收敛

这种现象称为 `Internal Covariate Shift`
> 训练集和测试集的分布不同，称为样本之间存在 `Covariate Shift`

- 如果网络非常深，每一层的激活值可能越来越分散
- 如果用sigmoid激活函数，越来越大的x值可能导致导数接近0，这就是梯度消失的原因
- 梯度消失会导致训练越来越缓慢而难以收敛
- 使用Relu激活并且合适的参数初始化和较小的学习率可以改善这个现象，但是数据仍有发散的可能性

## BN Algo
- 求出 mini-batch 的均值
- 求出 mini-batch 的方差
- Normalize 操作
  - 增加 ε ， 防止方差为0 
- 增加 γ , β
  - 让隐藏层之前有些微不同分布，增加非线性
  -  γ , β 和 w 一样进行梯度下降更新值

---
![image](https://user-images.githubusercontent.com/2216970/70209323-f1dd3600-176a-11ea-8b00-34e6f6c70527.png)

## Test Result
![image](https://user-images.githubusercontent.com/2216970/70210057-b2afe480-176c-11ea-8fc6-4875708ffe84.png)
![image](https://user-images.githubusercontent.com/2216970/70210125-d8d58480-176c-11ea-965b-5c15f7499754.png)

# Others

数据经过标准化后，其形状保持大致不变，但尺寸被我们压缩至（-1, 1）之间，而原尺寸在(-80，80)之间。
![image](https://user-images.githubusercontent.com/2216970/53677525-4b03cf00-3cec-11e9-8c00-fe640363b653.png)
![image](https://user-images.githubusercontent.com/2216970/53677528-6cfd5180-3cec-11e9-9a5c-d5c87a4a8aaa.png)

Accuracy | ![image](https://user-images.githubusercontent.com/2216970/53677530-7686b980-3cec-11e9-9801-5b3087f4a0f6.png) 
-- | -- 
Loss | ![image](https://user-images.githubusercontent.com/2216970/53677532-7981aa00-3cec-11e9-838d-e3b4065d5baf.png)

**Comment by junxnone** [comment]
# BatchNormalization Layer

**Comment by junxnone** [comment]
# ScaleShift Layer

ScaleShift layer performs linear transformation of the input blobs. Weights denote scaling parameter, biases - a shift.

**Comment by junxnone** [comment]
#823
